File: scripts/game_modes.py
# scripts/game_modes.py
# Разные режимы игры: сюжет, выживание, аркада, пазл и т.д.

import random
import time

class GameMode:
    """Базовый класс для любого режима игры"""
    def __init__(self, name="Default"):
        self.name = name
        self.start_time = time.time()
        self.score = 0
        self.is_active = True

    def update(self, player, dt, keys, game_state):
        """Вызывается каждый кадр"""
        pass

    def draw_ui(self, surface, font):
        """Рисует специфический интерфейс режима"""
        pass

    def on_event(self, event):
        """Обработка событий (например, нажатия)"""
        pass

    def end_mode(self):
        self.is_active = False
        print(f"Режим {self.name} завершён")


class StoryMode(GameMode):
    """Режим сюжета — использует storyline.py"""
    def __init__(self):
        super().__init__("Story")
        from scripts.storyline import story_manager
        self.story = story_manager
        self.current_dialog_timer = 0

    def update(self, player, dt, keys, game_state):
        # Прогресс по сюжету
        self.story.check_objective(player, game_state)

        # Показ диалога первые 5 секунд уровня
        if not self.story.current_dialog_shown:
            self.current_dialog_timer += dt
            if self.current_dialog_timer > 5:
                self.story.current_dialog_shown = True

    def draw_ui(self, surface, font):
        self.story.show_dialog(surface, font)


class SurvivalMode(GameMode):
    """Режим выживания — враги спавнятся бесконечно, цель — продержаться как можно дольше"""
    def __init__(self):
        super().__init__("Survival")
        self.spawn_timer = 0
        self.spawn_interval = 3.0  # секунды между спавнами
        self.enemies = []  # список врагов (пока заглушка)

    def update(self, player, dt, keys, game_state):
        self.spawn_timer += dt
        if self.spawn_timer >= self.spawn_interval:
            self.spawn_timer = 0
            # Спавн "врага" (пока просто увеличиваем счётчик)
            self.score += 1
            self.spawn_interval = max(1.0, self.spawn_interval - 0.1)  # ускоряем со временем

        # Пример: если игрок "умирает" — конец режима
        if player.lives <= 0:
            self.end_mode()

    def draw_ui(self, surface, font):
        time_survived = int(time.time() - self.start_time)
        surf = font.render(f"Время выживания: {time_survived} сек | Врагов: {self.score}", True, (255, 100, 100))
        surface.blit(surf, (20, 20))


class ArcadeMode(GameMode):
    """Аркадный режим — собирай очки, избегай препятствий, бесконечный"""
    def __init__(self):
        super().__init__("Arcade")
        self.coins = 0
        self.speed_multiplier = 1.0

    def update(self, player, dt, keys, game_state):
        # Пример: каждые 10 секунд ускоряем игру
        if int(time.time() - self.start_time) % 10 == 0 and int(time.time() - self.start_time) > 0:
            self.speed_multiplier += 0.2
            player.speed = 6 * self.speed_multiplier

        # Собирание "монет" (заглушка — можно добавить реальные объекты)
        if random.random() < 0.01:  # шанс "подобрать монету"
            self.coins += 1

    def draw_ui(self, surface, font):
        surf = font.render(f"Монеты: {self.coins} | Скорость: x{self.speed_multiplier:.1f}", True, (255, 215, 0))
        surface.blit(surf, (20, 20))


# Словарь режимов для быстрого выбора
MODES = {
    "story": StoryMode,
    "survival": SurvivalMode,
    "arcade": ArcadeMode,
    # Добавляй новые: "puzzle": PuzzleMode, "boss": BossRushMode и т.д.
}

def create_mode(mode_name):
    if mode_name in MODES:
        return MODES[mode_name]()
    return GameMode("Unknown")

File: js/game_modes.js
// js/game_modes.js
// Режимы игры — сюжет, выживание, аркада, пазл, босс-раш и т.д. (много режимов с разными правилами)

class GameMode {
  constructor(name) {
    this.name = name;
    this.score = 0;
    this.timer = 0;
    this.isActive = true;
  }

  update(player, dt, keys, gameState) {
    this.timer += dt;
  }

  drawUI(ctx) {
    ctx.fillStyle = '#ffffff';
    ctx.font = '20px Arial';
    ctx.fillText(`${this.name} режим`, 20, 60);
    ctx.fillText(`Очки: ${Math.floor(this.score)}`, 20, 90);
  }

  end() {
    this.isActive = false;
    soundManager.play('win', 1.0);
  }
}

class StoryMode extends GameMode {
  constructor() {
    super('Сюжет');
    this.story = storyManager;
  }

  update(player, dt, keys, gameState) {
    super.update(player, dt, keys, gameState);
    this.story.update(dt, player, gameState);
    // Пример: если цель уровня выполнена — награда и переход
    if (player.score >= 5 && this.story.currentLevel === 0) {
      unlockRandomPerk(player);
      this.story.advanceLevel(player);
      player.score = 0;
    }
  }

  drawUI(ctx) {
    super.drawUI(ctx);
    this.story.showDialog(ctx);
  }
}

class SurvivalMode extends GameMode {
  constructor() {
    super('Выживание');
    this.spawnTimer = 0;
    this.spawnInterval = 4.0;
  }

  update(player, dt, keys, gameState) {
    super.update(player, dt, keys, gameState);
    this.score += dt * 10;  // очки за время

    this.spawnTimer += dt;
    if (this.spawnTimer >= this.spawnInterval) {
      this.spawnTimer = 0;
      enemyManager.enemies.push(new Enemy(Math.random() * canvas.width, -50));
      this.spawnInterval = Math.max(1.5, this.spawnInterval - 0.2);
      soundManager.play('enemy_spawn', 0.6);
    }

    if (player.lives <= 0) this.end();
  }

  drawUI(ctx) {
    super.drawUI(ctx);
    ctx.fillText(`Время выживания: ${this.timer.toFixed(1)} сек`, 20, 120);
  }
}

class ArcadeMode extends GameMode {
  constructor() {
    super('Аркада');
    this.coinTimer = 0;
  }

  update(player, dt, keys, gameState) {
    super.update(player, dt, keys, gameState);
    this.score += dt * 5;

    this.coinTimer += dt;
    if (this.coinTimer >= 2.0) {
      this.coinTimer = 0;
      // Симулируем монету (можно добавить класс Coin позже)
      player.score += 10;
      soundManager.play('collect', 0.8);
      particleSystem.addExplosion(player.x + 32, player.y + 32, 20, '#ffff00');
    }

    if (this.timer > 60) this.score *= 1.5;  // бонус за минуту
  }

  drawUI(ctx) {
    super.drawUI(ctx);
    ctx.fillText(`Монеты: ${Math.floor(this.score / 10)}`, 20, 120);
  }
}

// Дополнительные режимы (можно расширять)
class PuzzleMode extends GameMode {
  constructor() {
    super('Пазл');
  }
  update(player, dt, keys, gameState) {
    // Логика пазла — например, собрать элементы в правильном порядке
    super.update(player, dt, keys, gameState);
  }
}

class BossRushMode extends GameMode {
  constructor() {
    super('Босс-раш');
    this.bossHealth = 100;
  }
  update(player, dt, keys, gameState) {
    super.update(player, dt, keys, gameState);
    // Босс появляется после 30 сек
    if (this.timer > 30) {
      // Симуляция урона боссу
      if (keys[' ']) this.bossHealth -= dt * 20;
      if (this.bossHealth <= 0) this.end();
    }
  }
  drawUI(ctx) {
    super.drawUI(ctx);
    ctx.fillText(`Босс HP: ${Math.max(0, this.bossHealth)}`, 20, 120);
  }
}

// Словарь режимов
const MODE_FACTORY = {
  story: () => new StoryMode(),
  survival: () => new SurvivalMode(),
  arcade: () => new ArcadeMode(),
  puzzle: () => new PuzzleMode(),
  bossrush: () => new BossRushMode()
  // Добавь ещё: co-op симуляция, speedrun, endless, defense и т.д.
};

function createMode(name) {
  if (MODE_FACTORY[name]) {
    const mode = MODE_FACTORY[name]();
    soundManager.play('mode_start', 0.9);
    return mode;
  }
  return new GameMode('Неизвестный');
}
